### PR TITLE
DROOLS-4042: [DMN Designer] Add support for importing and consuming PMML models

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-4042

Depends on https://github.com/kiegroup/drools/pull/2394

This PR adds support to import PMML documents with the Included Models tab. PMML documents are _discovered_ the same as DMN models (i.e. in `-standalone` they're in the `diagrams` folder in the underlying Git repository and in Business Central they can be in the same project). Once imported they can be used in a `Function` Boxed Expression that has `Kind` of `PMML`.

This PR adds the following:
- Inclusion of DMN and PMML files in the Include Model dialog. 
- New Included Model _card_ for PMML that includes the PMML model count in the PMML document.
- Use of list boxes in the `Function` Boxed Expression when the `Kind` is `PMML`.
- Automatic creation of `Function` parameters for the PMML document/model.
- `locationURI` of include models are now _relative_ to the file importing the model.

**Dialog changes**
DROOLS-4042-include-dialog

**Included Model cards**
DROOLS-4042-includes

**Function Boxed Expression editor**
DROOLS-4042-function-pmml

**Example DMN XML imports section**
```
<dmn:import id="_2C36DB08-80F2-4C48-BA79-193C25706FF2" name="dmn-import" namespace="https://kiegroup.org/dmn/_8AEEC670-6D6D-479D-92AE-B9A97578995B" locationURI="EpAmKkh7.dmn" importType="http://www.omg.org/spec/DMN/20180521/MODEL/"></dmn:import>

<dmn:import id="_A9803527-9127-4F38-A27E-48730889A2E1" name="pmml-import" namespace="pmml-import" locationURI="test_scorecard.pmml" importType="http://www.dmg.org/PMML-4_2"></dmn:import>
```

### CODE REVIEW

**Comment by manstis** [comment]
@karreiro Following our discussion on GChat I opened a [PR](https://github.com/kiegroup/drools-wb/pull/1166) for `drools-wb` too.

**Comment by manstis** [comment]
@jomarko @karreiro This PR is "good to go" :-)

**Comment by karreiro** [comment]
@manstis do we need the PR on [`droolsjbpm-knowledge`](https://github.com/manstis/droolsjbpm-knowledge/commit/102e981de62c1dbb069abf494c03313afb8a1845)?

**Comment by manstis** [comment]
@karreiro (and @jomarko) no; just this PR.. but the drools artifacts that had their PRs merged this morning (GMT) may not have been published as new SNAPSHOTs yet. If you're desperate to look at this today you'll need to fetch {{master}} of {{droolsjbpm-build-bootstrap}} and {{drools}} and build them locally.

**Comment by karreiro** [comment]
Thank you, @manstis :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/DMNClient.gwt.xml [hunk 1]
```diff
@@ -26,6 +26,7 @@
   <inherits name="org.kie.workbench.common.stunner.client.StunnerWidgets"/>
   <inherits name="org.kie.workbench.common.stunner.forms.StunnerFormsClient"/>
   <inherits name="org.kie.workbench.common.widgets.KieWorkbenchWidgetsCommon"/>
+  <inherits name="org.guvnor.common.services.project.GuvnorProjectClient"/>
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/DMNClient.gwt.xml]: @manstis Do we need this line? (I don't think so, but maybe I'm missing something).

**Comment by jomarko** [comment]
### Manual review note 01

When user include sample pmml models from [documentation](https://docs.jboss.org/drools/release/latest/drools-docs/html_single/index.html#pmml-examples-ref_pmml-models), the shown count of contained (models?/elements?) is  zero as shown in picture

Screenshot from 2019-06-10 08-05-16

**Comment by jomarko** [comment]
### Manual review note 02

Function parameters have always type **Any** regardless of the real type in pmml, is it expected?

Screenshot from 2019-06-10 08-20-03

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelperImpl.java [hunk 2]
```diff
@@ -246,6 +277,10 @@ private void setInformationItem(final DRGElement drgElement,
                 .collect(Collectors.toList());
     }
 
+    List<Path> getPMMLDocumentPaths(final Metadata metadata) {
+        return pathsHelper.getPMMLModelsPaths(getProject(metadata));
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelperImpl.java]: This has no coverage, please fix.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/editors/common/PMMLIncludedDocumentFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import static java.util.Comparator.naturalOrder;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PMMLIncludedDocumentFactoryTest {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/editors/common/PMMLIncludedDocumentFactoryTest.java]: `PMMLIncludedDocumentFactory` coverage is still just 37% are we sure it is enough?

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/AllModelsValueFileExtensionIndexTerm.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+package org.kie.workbench.common.dmn.backend.editors.included.query;
+
+import org.kie.workbench.common.dmn.backend.editors.types.query.DMNValueFileExtensionIndexTerm;
+
+public class AllModelsValueFileExtensionIndexTerm extends DMNValueFileExtensionIndexTerm {
+
+    @Override
+    public String getValue() {
+        return ".*(dmn|pmml)";
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/AllModelsValueFileExtensionIndexTerm.java]: Please Test.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/PMMLValueFileExtensionIndexTerm.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+package org.kie.workbench.common.dmn.backend.editors.included.query;
+
+import org.kie.workbench.common.dmn.backend.editors.types.query.DMNValueFileExtensionIndexTerm;
+
+public class PMMLValueFileExtensionIndexTerm extends DMNValueFileExtensionIndexTerm {
+
+    @Override
+    public String getValue() {
+        return ".*(pmml)";
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/PMMLValueFileExtensionIndexTerm.java]: Please test

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/pmml/util/PMMLValueUtils.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.workbench.common.dmn.client.editors.expressions.types.function.supplementary.pmml.util;
+
+public class PMMLValueUtils {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/pmml/util/PMMLValueUtils.java]: Seems same logic as `org.kie.workbench.common.stunner.bpmn.client.forms.util.StringUtils`

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelperImpl.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelperImpl.java]: Test added (in new commit to follow...)

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/editors/common/PMMLIncludedDocumentFactoryTest.java (hunk 3 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/editors/common/PMMLIncludedDocumentFactoryTest.java]: More tests added (in new commit to follow). Coverage reported as 92%.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java [hunk 7]
```diff
@@ -61,31 +64,68 @@ public DMNPathsHelperImpl() {
...
+        if (Objects.isNull(dmnModelPath)) {
+            return includedModelPath.getFileName();
+        }
+
+        final org.uberfire.java.nio.file.Path nioDMN = convertPath(normalizePath(dmnModelPath));
+        final org.uberfire.java.nio.file.Path nioIncluded = convertPath(normalizePath(includedModelPath));
+        //This can return Path like "../file" for files in the same folder so it needs to be normalised.
+        final org.uberfire.java.nio.file.Path nioRelative = nioDMN.relativize(nioIncluded);
+
+        //Path.normalise() has a bug for Paths prefixed "../" so strip it manually. toString() works better than toURI().toString()!
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java]: Maybe link the bug if reported somewhere? Would be good as soon as possible get rid of the `commons-io` dependency.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/AllModelsValueFileExtensionIndexTerm.java (hunk 4 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/AllModelsValueFileExtensionIndexTerm.java]: Test added in new commit to follow..

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/PMMLValueFileExtensionIndexTerm.java (hunk 5 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/query/PMMLValueFileExtensionIndexTerm.java]: Test added in new commit to follow..

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/definition/v1_1/ImportConverter.java [hunk 8]
```diff
@@ -17,25 +17,31 @@
...
+import org.kie.workbench.common.dmn.api.definition.v1_1.ImportDMN;
+import org.kie.workbench.common.dmn.api.definition.v1_1.ImportPMML;
+import org.kie.workbench.common.dmn.api.editors.included.DMNImportTypes;
+import org.kie.workbench.common.dmn.api.editors.included.PMMLDocumentMetadata;
 import org.kie.workbench.common.dmn.api.property.dmn.Id;
 import org.kie.workbench.common.dmn.api.property.dmn.LocationURI;
 import org.kie.workbench.common.dmn.api.property.dmn.Name;
 import org.kie.workbench.common.dmn.api.property.dmn.QName;
 
+import static org.kie.workbench.common.dmn.api.editors.included.DMNImportTypes.determineImportTypeType;
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/definition/v1_1/ImportConverter.java]: Is the name intentional? I mean ending with `TypeType`

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java [hunk 9]
```diff
@@ -2219,6 +2231,45 @@ public void testConnectorCentreToCentre() throws Exception {
...
+        assertTrue(imports.get(1) instanceof ImportPMML);
+
+        final ImportDMN dmnImport = (ImportDMN) imports.get(0);
+        assertEquals("dmn-import", dmnImport.getName().getValue());
+        assertEquals("https://kiegroup.org/dmn/_46EB0D0D-7241-4629-A38E-0377AA5B32D1", dmnImport.getNamespace());
+        assertEquals(DMNImportTypes.DMN.getDefaultNamespace(), dmnImport.getImportType());
+
+        final ImportPMML pmmlImport = (ImportPMML) imports.get(1);
+        assertEquals("pmml-import", pmmlImport.getName().getValue());
+        assertEquals("pmml-import", pmmlImport.getNamespace());
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java]: Is this correct? Shouldn't be something more like: "...dmg.org...."?

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImplTest.java [hunk 10]
```diff
@@ -160,14 +227,73 @@ public void testGetStandaloneRootPath() {
...
+        when(includedModelPath.getFileName()).thenReturn("file1");
+
+        assertEquals("file1", pathsHelper.getRelativeURI(null, includedModelPath));
+    }
+
+    @Test
+    public void testGetRelativeURIWhenSameLevel() {
+        doTestGetRelativeURI("dmnModel.dmn",
+                             "includedModel.dmn",
+                             "../includedModel.dmn",
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImplTest.java]: Is this correct as test method says same level?

**Comment by jomarko** [comment]
The error below says need to fill both **Document** and **Model**
Screenshot from 2019-06-10 12-33-24


However the error is gone while **Model** Still mising
Screenshot from 2019-06-10 12-33-33

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/pmml/util/PMMLValueUtils.java (hunk 6 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/pmml/util/PMMLValueUtils.java]: I have moved the applicable methods from BPMNs `StringUtils` to Stunners generic `StringUtils` and deleted DMNs `PMMLValueUtils` in the new commit.. I'll look at the `-api` and `-backend` comments on a separate commit (to hopefully make re-review easier).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/DMNClient.gwt.xml (hunk 1 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/DMNClient.gwt.xml]: @karreiro I too could not see why it was needed; so it has been removed in a new commit.

File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/editors/included/DMNImportTypes.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    public static DMNImportTypes determineImportTypeType(final String namespace) {
+        return importTypes
+                .stream()
+                .filter(importType -> importType.matchesNamespace(namespace))
+                .findFirst()
+                .orElse(null);
+    }
+
+    private boolean matchesNamespace(final String namespace) {
+        return namespaces.stream().anyMatch(ns -> ns.equals(namespace));
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/editors/included/DMNImportTypes.java]: Updated in new commit.

File: kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/editors/included/DMNImportTypes.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    public String getFileExtension() {
+        return fileExtension;
+    }
+
+    public String getDefaultNamespace() {
+        return namespaces.get(namespaces.size() - 1);
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/editors/included/DMNImportTypes.java]: Tests added for both methods. It is impossible to create another `DMNImportType` as both the constructor and `registerImportType(..)` method are `private` hence we can be confident there is at least always a single `namespace`.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/definition/v1_1/ImportConverter.java (hunk 8 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/definition/v1_1/ImportConverter.java]: Corrected to a single `Type` i.e. `determineImportType(..)`

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java (hunk 9 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java]: No, PMML documents do not have a `namespace` attribute themselves (like DMN models do) so the `name` is also used as the `namespace` (The DMN `Import` node must have a `namespace` attribute so we use the same value). This was proposed by @tarilabs (I don't think there's anything in the DMN specification stating this and, AFAIK, it was an internal decision).

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImplTest.java (hunk 10 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImplTest.java]: Yes, as `org.uberfire.java.nio.file.Path.relativize(..)` returns `../abc` when two files are in the same folder as commented [here](https://github.com/kiegroup/kie-wb-common/pull/2683/files/a4ec9ce74edd70dd9f6b9b805ac0c78622be6f49#diff-10d9bf93e36fa338010e0315f1347070R111).

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java (hunk 9 above)
Comment (tarilabs) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java]: We will seek clarification from RTF, but until then we deal with the best interpretation from the DMN spec

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/function/SetParametersCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        //Add parameter and then undo
+        assertEquals(GraphCommandResultBuilder.SUCCESS,
+                     c.execute(gce));
+        assertEquals(GraphCommandResultBuilder.SUCCESS,
+                     c.undo(gce));
+
+        assertFormalParameters();
+    }
+
+    private void assertFormalParameters(final InformationItem... parameters) {
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/function/SetParametersCommandTest.java]: Updated in new commit.

**Comment by manstis** [comment]
@jomarko Regarding:

> Manual review note 02

This is correct, for the time being we are mapping all PMML data-types to DMN's "Any".

**Comment by manstis** [comment]
@jomarko The issue relating to no error if only the PMML document is given (and not PMML model) appears to be a bug in `DMNEvaluatorCompiler.compileFunctionDefinitionPMML(..)`. Could you please report a DROOLS bug for @tarilabs to investigate.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java (hunk 7 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java]: See https://issues.jboss.org/browse/AF-2045

I have also managed to remove the dependency on `commons-io`.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
@jomarko Regarding:

> Manual review note 01

Please see https://issues.jboss.org/browse/DROOLS-4154 (it is an engine issue).

**Comment by manstis** [comment]
Jenkins please retest this.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java (hunk 7 above)
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java]: Could we please mention that ticket number in code/comment?

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNPathsHelperImpl.java]: I do. Please look at the file at the latest commit.

**Comment by jomarko** [comment]
Outstanding issue that will be solved separately:
- https://issues.jboss.org/browse/DROOLS-4155
- https://issues.jboss.org/browse/DROOLS-4156

**Comment by manstis** [comment]
@jomarko I think this PR is ready for re-review (I was looking into https://issues.jboss.org/browse/DROOLS-4155 last night, before you filed the JIRA however now it's being tracked separately I think this PR can continue its journey....)

**Comment by manstis** [comment]
Jenkins please execute full downstream build.

**Comment by manstis** [comment]
Jenkins please execute full downstream build.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Pull request merged**